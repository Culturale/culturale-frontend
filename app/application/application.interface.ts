import type { IRequestSubject } from '~/observables';

import type { Controllers } from './controllers';
import type { UseCasesMap } from './use-cases/use-cases';

export type UseCaseRequestID = `${keyof UseCasesMap}-${string}` | keyof UseCasesMap;

/**
 * Abstraction of the application layer of the application
 * @param routeController -> Module that handles the route assigned to the driver
 * @param locationController -> Module that handles the device's location operations
 * @param navigationController -> Module that handles the navigation operations
 */
export interface IApplication {
  readonly controllers: Controllers;
  readonly useCases: UseCasesMap;

  readonly controllerErrors: Map<string, Error>;
  readonly useCasesRequests: Map<UseCaseRequestID, IRequestSubject<unknown>>;

  setup: () => Promise<void>;
}
