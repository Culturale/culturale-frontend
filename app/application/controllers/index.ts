export * from './route-controller';
export * from './driver-controller';
export * from './controller';
export * from './controller.interface';
