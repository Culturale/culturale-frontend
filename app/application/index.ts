export * from './application';
export * from './application.mock';
export * from './application.interface';
