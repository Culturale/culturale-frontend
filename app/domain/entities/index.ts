export * from './chat';
export * from './event';
export * from './message';
export * from './user';