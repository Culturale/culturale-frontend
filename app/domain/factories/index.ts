export * from './event.factory';
export * from './user.factory';
export * from './message.factory';
export * from './chat.factory';
