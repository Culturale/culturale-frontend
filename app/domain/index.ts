export * from './entities';
export * from './factories';
export * from './repositories';
