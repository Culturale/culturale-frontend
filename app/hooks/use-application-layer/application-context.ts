import { createContext, useContext } from 'react';

import type { IApplication } from '~/application';

const ApplicationContext = createContext<IApplication>({} as IApplication);

export const ApplicationLayerProvider = ApplicationContext.Provider;

export const useApplicationLayer = () => useContext(ApplicationContext);
