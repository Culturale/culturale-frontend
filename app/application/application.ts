import { computed, makeObservable, observable } from 'mobx';

import type { IInfrastructure } from '~/infrastructure';
import { Infrastructure } from '~/infrastructure';
import type { IRequestSubject } from '~/observables';

import type { IApplication, UseCaseRequestID } from './application.interface';
import type { Controllers } from './controllers';
import { EventController, UserController } from './controllers';
import { login } from './use-cases';
import { signup } from './use-cases/signup';
import type { UseCasesMap } from './use-cases/use-cases';
import { newmessage } from './use-cases/newmessage';
import { fetchPaymentSheetParams } from './use-cases/fetch-payment-sheet-params';

export class Application implements IApplication {
  public readonly controllers: Controllers;

  public readonly useCases: UseCasesMap;

  private infrastructure: IInfrastructure;
  private errors: Map<string, Error> = new Map<string, Error>();
  public readonly useCasesRequests: Map<
    UseCaseRequestID,
    IRequestSubject<unknown>
  > = new Map<UseCaseRequestID, IRequestSubject<unknown>>();

  constructor() {
    this.infrastructure = new Infrastructure();

    this.controllers = {
      EventController: new EventController(this.infrastructure),
      UserController: new UserController(this.infrastructure),
    };

    this.useCases = {
      Login: (username: string, password: string) => {
        const subject = login(
          this.infrastructure,
          this.controllers.UserController,
          username,
          password,
        );
        this.useCasesRequests.set('Login', subject);
        return subject;
      },
      FetchPaymentSheetParams: (eventId: string) => {
        return fetchPaymentSheetParams(this.infrastructure, eventId);
      },
      Signup: (
        username: string,
        name: string,
        password: string,
        email: string,
        phoneNumber: string,
        userType: string,
        profilePicture?: string,
      ) => {
        const subject = signup(
          this.infrastructure,
          this.controllers.UserController,
          username,
          name,
          password,
          email,
          phoneNumber,
          userType,
          profilePicture,
        );
        this.useCasesRequests.set('Signup', subject);
        return subject;
      },
      NewMessage: (id: string, content: string, userId: string) => {
        const date = new Date();
        const subject = newmessage(
          this.infrastructure,
          id,
          content,
          userId,
          date,
        );
        this.useCasesRequests.set('NewMessage', subject);
        return subject;
      },
    };

    makeObservable(this, {
      controllerErrors: computed,
      useCasesRequests: observable,
    });
  }

  public async setup(): Promise<void> {
    await this.controllers.UserController.setup();
    if (this.controllers.UserController.token) {
      this.infrastructure.api.setup(this.controllers.UserController.token);
    }
  }

  public get controllerErrors(): Map<string, Error> {
    return this.errors;
  }
}
