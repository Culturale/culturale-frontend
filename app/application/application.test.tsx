import { observer } from 'mobx-react-lite';
import * as React from 'react';
import { Button, View } from 'react-native';

import { Route } from '~/domain';
import { useApplicationLayer } from '~/hooks/use-application-layer/application-context';
import type {
  RenderAPI} from '~/test/utils';
import {
  renderWithApplicationLayer,
  fireEvent,
  waitFor
} from '~/test/utils';

import { Application } from './application';
import type { IApplication } from './application.interface';
import { MockApplication } from './application.mock';

jest.mock('jwt-decode', () => jest.fn());

describe('Application Layer', function () {
  let instance: IApplication;

  beforeEach(function () {
    jest.useFakeTimers();
    instance = new Application();
  });

  afterAll(function () {
    jest.useRealTimers();
  });

  it('can be created', function () {
    expect(instance).toMatchSnapshot();
  });

  // Controller Errors Handler Test (To implement with new controllers)
  /*   describe('handles controllers errors correctly', function () {
    const requestKey = LocationRequests.UpdateCurrentLocation;

    beforeEach(function () {
      const updateCurrentLocationSubject =
        instance.controllers.LocationController.requestsSubjectsMap[requestKey];

      updateCurrentLocationSubject.failRequest(new Error('test-error'));
    });

    it('sets the error to the error map with the request ID as key', function () {
      expect(instance.controllerErrors.get(requestKey)).toStrictEqual(
        new Error('test-error')
      );
    });

    it('after an error is tracked, the subject is able to report another error', function () {
      const updateCurrentLocationSubject2 =
        instance.controllers.LocationController.requestsSubjectsMap[requestKey];
      updateCurrentLocationSubject2.failRequest(new Error('test-error-2'));

      expect(instance.controllerErrors.get(requestKey)).toStrictEqual(
        new Error('test-error-2')
      );
    });

    it('after a request is completed, it can track again the errors', function () {
      const updateCurrentLocationSubject2 =
        instance.controllers.LocationController.requestsSubjectsMap[requestKey];
      updateCurrentLocationSubject2.completeRequest();

      const updateCurrentLocationSubject3 =
        instance.controllers.LocationController.requestsSubjectsMap[requestKey];
      updateCurrentLocationSubject3.failRequest(new Error('test-error-2'));

      expect(instance.controllerErrors.get(requestKey)).toStrictEqual(
        new Error('test-error-2')
      );
    });
  }); */

  describe('is observable', function () {
    let instance: RenderAPI;
    const reRenderHandler = jest.fn();
    let app: IApplication;

    function renderInstance(application: IApplication) {
      const Component: React.FC = observer(() => {
        const {
          controllers: { RouteController }
        } = useApplicationLayer();
        const route = RouteController.route;

        React.useEffect(() => reRenderHandler(), [route]);

        return (
          <View>
            <Button
              title='Update State'
              onPress={() => RouteController.setRoute(new Route())}
            />
          </View>
        );
      });

      return renderWithApplicationLayer(<Component />, application);
    }

    beforeEach(function () {
      app = new Application();
      instance = renderInstance(app);
      expect(reRenderHandler).toHaveBeenCalledTimes(1);
    });

    it('calls reRenderHandler when RouteController sets a new value at route', function () {
      const button = instance.getByText(/update state/i);
      fireEvent(button, 'press');

      expect(reRenderHandler).toHaveBeenCalledTimes(2);
    });
  });

  describe('you can subscribe to use cases requests', function () {
    let instance: RenderAPI;
    const reRenderHandler = jest.fn();
    const handleRequestCompleted = jest.fn();
    let app: IApplication;

    function renderInstance(application: IApplication) {
      const Component: React.FC = observer(() => {
        const {
          useCasesRequests,
          useCases: { Login }
        } = useApplicationLayer();

        const loginRequest = useCasesRequests.get('Login');

        React.useEffect(() => {
          reRenderHandler();
          loginRequest?.subscribeToRequest({
            onCompleteRequest: handleRequestCompleted
          });
        }, [loginRequest]);

        return (
          <View>
            <Button title='Log In' onPress={() => Login()} />
          </View>
        );
      });

      return renderWithApplicationLayer(<Component />, application);
    }

    beforeEach(function () {
      app = new MockApplication();
      instance = renderInstance(app);
    });

    it('when a use case is triggered, the use cases requests map is updated', function () {
      const button = instance.getByText(/log in/i);
      fireEvent(button, 'press');

      expect(reRenderHandler).toHaveBeenCalledTimes(2);
    });

    it('you can subscribe to a subject of the use cases requests map', async function () {
      jest.useRealTimers();
      handleRequestCompleted.mockClear();
      const button = instance.getByText(/log in/i);
      fireEvent(button, 'press');

      await waitFor(() => expect(handleRequestCompleted).toBeCalled());
    });

    it('you can subscribe to a subject with unique ID of the use cases requests map', async function () {
      jest.useRealTimers();
      handleRequestCompleted.mockClear();
      const button = instance.getByText(/log in/i);
      fireEvent(button, 'press');

      await waitFor(() => expect(handleRequestCompleted).toBeCalled());
    });
  });
});
