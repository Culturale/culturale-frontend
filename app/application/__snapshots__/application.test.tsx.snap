// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application Layer can be created 1`] = `
Application {
  "controllers": Object {
    "DriverController": DriverController {
      "driverID": undefined,
      "fleetEngineAccessToken": undefined,
      "isLoggedIn": null,
      "requestsSubjectsMap": Object {
        "LOGIN": RequestSubject {
          "_value": "IDLE",
          "closed": false,
          "currentObservers": null,
          "hasCompleted": false,
          "hasError": false,
          "hasFailed": false,
          "isStopped": false,
          "observers": Array [
            SafeSubscriber {
              "_finalizers": Array [
                Subscription {
                  "_finalizers": null,
                  "_parentage": [Circular],
                  "closed": false,
                  "initialTeardown": [Function],
                },
              ],
              "_parentage": null,
              "closed": false,
              "destination": ConsumerObserver {
                "partialObserver": Object {
                  "complete": [Function],
                  "error": [Function],
                  "next": undefined,
                },
              },
              "initialTeardown": undefined,
              "isStopped": false,
            },
          ],
          "requestID": "LOGIN",
          "requestStatus": "IDLE",
          "thrownError": null,
        },
        "LOGOUT": RequestSubject {
          "_value": "IDLE",
          "closed": false,
          "currentObservers": null,
          "hasCompleted": false,
          "hasError": false,
          "hasFailed": false,
          "isStopped": false,
          "observers": Array [
            SafeSubscriber {
              "_finalizers": Array [
                Subscription {
                  "_finalizers": null,
                  "_parentage": [Circular],
                  "closed": false,
                  "initialTeardown": [Function],
                },
              ],
              "_parentage": null,
              "closed": false,
              "destination": ConsumerObserver {
                "partialObserver": Object {
                  "complete": [Function],
                  "error": [Function],
                  "next": undefined,
                },
              },
              "initialTeardown": undefined,
              "isStopped": false,
            },
          ],
          "requestID": "LOGOUT",
          "requestStatus": "IDLE",
          "thrownError": null,
        },
        "TOTAL_CASH": RequestSubject {
          "_value": "IDLE",
          "closed": false,
          "currentObservers": null,
          "hasCompleted": false,
          "hasError": false,
          "hasFailed": false,
          "isStopped": false,
          "observers": Array [
            SafeSubscriber {
              "_finalizers": Array [
                Subscription {
                  "_finalizers": null,
                  "_parentage": [Circular],
                  "closed": false,
                  "initialTeardown": [Function],
                },
              ],
              "_parentage": null,
              "closed": false,
              "destination": ConsumerObserver {
                "partialObserver": Object {
                  "complete": [Function],
                  "error": [Function],
                  "next": undefined,
                },
              },
              "initialTeardown": undefined,
              "isStopped": false,
            },
          ],
          "requestID": "TOTAL_CASH",
          "requestStatus": "IDLE",
          "thrownError": null,
        },
        "USER_INFO": RequestSubject {
          "_value": "IDLE",
          "closed": false,
          "currentObservers": null,
          "hasCompleted": false,
          "hasError": false,
          "hasFailed": false,
          "isStopped": false,
          "observers": Array [
            SafeSubscriber {
              "_finalizers": Array [
                Subscription {
                  "_finalizers": null,
                  "_parentage": [Circular],
                  "closed": false,
                  "initialTeardown": [Function],
                },
              ],
              "_parentage": null,
              "closed": false,
              "destination": ConsumerObserver {
                "partialObserver": Object {
                  "complete": [Function],
                  "error": [Function],
                  "next": undefined,
                },
              },
              "initialTeardown": undefined,
              "isStopped": false,
            },
          ],
          "requestID": "USER_INFO",
          "requestStatus": "IDLE",
          "thrownError": null,
        },
      },
      "token": undefined,
      "totalCash": undefined,
      "userInfo": undefined,
    },
    "RouteController": RouteController {
      "driverId": undefined,
      "infrastructure": Infrastructure {
        "api": API {
          "axiosClient": undefined,
          "baseURL": "https://test-api.com",
          "token": undefined,
        },
        "services": Object {
          "location": LocationService {
            "lastKnownLocation": null,
            "locationUpdatesSubscription": null,
            "options": Object {},
          },
        },
      },
      "requestsSubjectsMap": Object {
        "REQUEST_ROUTE_DOCUMENT_ID": RequestSubject {
          "_value": "IDLE",
          "closed": false,
          "currentObservers": null,
          "hasCompleted": false,
          "hasError": false,
          "hasFailed": false,
          "isStopped": false,
          "observers": Array [
            SafeSubscriber {
              "_finalizers": Array [
                Subscription {
                  "_finalizers": null,
                  "_parentage": [Circular],
                  "closed": false,
                  "initialTeardown": [Function],
                },
              ],
              "_parentage": null,
              "closed": false,
              "destination": ConsumerObserver {
                "partialObserver": Object {
                  "complete": [Function],
                  "error": [Function],
                  "next": undefined,
                },
              },
              "initialTeardown": undefined,
              "isStopped": false,
            },
          ],
          "requestID": "REQUEST_ROUTE_DOCUMENT_ID",
          "requestStatus": "IDLE",
          "thrownError": null,
        },
      },
      "route": Route {
        "route": Array [],
        "routeState": "FULLY_SEQUENCED",
      },
      "routeId": undefined,
      "setDriverId": undefined,
    },
  },
  "errorSubscription": [Function],
  "errors": Map {},
  "infrastructure": Infrastructure {
    "api": API {
      "axiosClient": undefined,
      "baseURL": "https://test-api.com",
      "token": undefined,
    },
    "services": Object {
      "location": LocationService {
        "lastKnownLocation": null,
        "locationUpdatesSubscription": null,
        "options": Object {},
      },
    },
  },
  "useCases": Object {
    "Login": [Function],
  },
  "useCasesRequests": Array [],
}
`;
